core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      }
+
+      return this;
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
+      this.changes.add(new MetadataUpdate.AddViewVersion(version));
+      this.history.add(
+          ImmutableViewHistoryEntry.builder()
+              .timestampMillis(version.timestampMillis())
+              .versionId(version.versionId())
+              .build());
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> updated) {
+      if (updated.isEmpty()) {
+        return this;
+      }
+
+      this.properties.putAll(updated);
+      this.changes.add(new MetadataUpdate.SetProperties(updated));
+      return this;
+    }
+
+    public Builder removeProperties(Set<String> propertiesToRemove) {
+      if (propertiesToRemove.isEmpty()) {
+        return this;
+      }
+
+      propertiesToRemove.forEach(this.properties::remove);
+      this.changes.add(new MetadataUpdate.RemoveProperties(propertiesToRemove));
+      return this;
+    }
+
+    private static Map<Integer, ViewVersion> indexVersions(List<ViewVersion> versionsToIndex) {
+      ImmutableMap.Builder<Integer, ViewVersion> builder = ImmutableMap.builder();
+      for (ViewVersion version : versionsToIndex) {
+        builder.put(version.versionId(), version);
+      }
+
+      return builder.build();
+    }
+
+    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
+      ImmutableMap.Builder<Integer, Schema> builder = ImmutableMap.builder();
+      for (Schema schema : schemasToIndex) {
+        builder.put(schema.schemaId(), schema);
+      }
+
+      return builder.build();
+    }
+
+    public ViewMetadata build() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
+      Preconditions.checkArgument(history.size() > 0, "Invalid view history: empty");
+      Preconditions.checkArgument(schemas.size() > 0, "Invalid schemas: empty");
+
+      Map<Integer, ViewVersion> versionsById = indexVersions(versions);
+      Preconditions.checkArgument(
+          versionsById.containsKey(currentVersionId),
+          "Cannot find current version %s in view versions: %s",
+          currentVersionId,
+          versionsById.keySet());
+
+      int currentSchemaId = versionsById.get(currentVersionId).schemaId();
+      Map<Integer, Schema> schemasById = indexSchemas(schemas);
+      Preconditions.checkArgument(
+          schemasById.containsKey(currentSchemaId),
+          "Cannot find current schema with id %s in schemas: %s",
+          currentSchemaId,
+          schemasById.keySet());
+
+      int versionHistorySizeToKeep =
+          PropertyUtil.propertyAsInt(
+              properties,
+              ViewProperties.VERSION_HISTORY_SIZE,
+              ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+
+      if (versionHistorySizeToKeep <= 0) {
```

Comment:
Now that this is in the builder, we can throw an exception if it is invalid. Maybe convert this to a precondition?

Reply:
makes sense, added a precondition and a test to ensure this fails properly